src/Microsoft.AspNet.OData/Extensions/HttpRequestMessageExtensions.cs

```diff
@@ -410,6 +412,57 @@ public static IEnumerable<IODataRoutingConvention> GetRoutingConventions(this Ht
 
             return request.GetRequestContainer().GetServices<IODataRoutingConvention>();
         }
+        /// <summary>
+        /// Checks whether the request is a POST targeted at a resource path ending in /$query.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="oDataPath">The OData path.</param>
+        /// <returns>true if the request path has $query segment.</returns>
+        internal static bool IsQueryRequest(this HttpRequestMessage request, string oDataPath)
+        {
+            return request.Method.Equals(HttpMethod.Post) && 
+                oDataPath?.TrimEnd('/').EndsWith('/' + ODataRouteConstants.QuerySegment, StringComparison.OrdinalIgnoreCase) == true;
+        }
+
+        /// <summary>
+        /// Transforms a POST request targeted at a resource path ending in $query into a GET request. 
+        /// The query options are parsed from the request body and appended to the request URL.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="requestContainerFactory">Request container factory.</param>
+        internal static void TransformQueryRequest(this HttpRequestMessage request, Func<IServiceProvider> requestContainerFactory)
```

Comment:
>, Func<IServiceProvider> requestContainerFactory [](start = 82, length = 48)

why pass in requestContainerFactory rather than using request.GetRequestContainer() as is done in the other extension methods?

Reply:
@mikepizzo If I call `request.GetRequestContainer()` at this point, the method will in turn call `request.CreateRequestContainer(request.ODataFeature().RuoteName)` and the problem will be that `RouteName` will be `null` at that point. We initialize the `Path` and `RouteName` properties of `request.ODataFeature()` [much further below](https://github.com/OData/WebApi/blob/d02bc61ea7b31ada1e54abbeebbecb3c5df0e3ac/src/Microsoft.AspNetCore.OData/Routing/ODataPathRouteConstraint.cs#L60) after the call to `GetODataPath()`.
Things are made worse by the fact that there's a call to `request.CreateRequestContainer(RouteName)` on the `GetODataPath()` [method call](https://github.com/OData/WebApi/blob/d02bc61ea7b31ada1e54abbeebbecb3c5df0e3ac/src/Microsoft.AspNetCore.OData/Routing/ODataPathRouteConstraint.cs#L57). The `CreateRequestContainer()` method is designed in such a way that it throws an exception if a request container already exists.
I therefore opted to add [this delegate](https://github.com/OData/WebApi/pull/2191/files#diff-a08977e326b20ea2c1a30644a20643beR51) that both `TransformQueryRequest()` and `GetODataPath()` methods invoke. The delegate won't blow up if the request container already exists.

Reply:
Fixed